### ISSUE TITLE
quadratic behaviour in Module::Scandeps::_add_info on case insensitive file systems

### ISSUE THREAD

**Issue by shawnlaffan** [rschupp/Module-ScanDeps#21]
Module::Scandeps::_add_info does a linear scan across the keys of the ```$rv``` hash when on a case insensitive file system (code block below). 

This leads to quadratic behaviour as the size of ```$rv``` increases.  Most of the searches are for ```$module``` or ```$used_by``` entries that have already been added to the hash so much of this is unnecessary.  

A simple solution is to check if the key exists before entering either of the loops.  This reduces the time under profiling from 309s to 77s.

Further, smaller, speed gains can be achieved by:
1. Using the same approach in the grep condition at https://github.com/rschupp/Module-ScanDeps/blob/c9cc39da7ebeab806a28cf052f365ecc2a47214a/lib/Module/ScanDeps.pm#L1189
2. Lifting the ```lc($module)``` and ```lc($used_by)``` calls to lexical vars outside the loops and then using these in the conditions.  
3. Abstracting all the ```File::Spec->case_tolerant()``` calls to a package lexical or constant.  

If that all seems sensible then I can work up a PR.

Of course, this might all be avoidable by filtering what is passed down to _add_deps?

Code block:

https://github.com/rschupp/Module-ScanDeps/blob/c9cc39da7ebeab806a28cf052f365ecc2a47214a/lib/Module/ScanDeps.pm#L1159-L1178

**Comment by shawnlaffan** [comment]
> Abstracting all the File::Spec->case_tolerant() calls to a package lexical or constant.

Although I now see there is already code to detect if the file system is case insensitive so the ```is_insensitive_fs``` constant could be used in more places.

https://github.com/rschupp/Module-ScanDeps/blob/c9cc39da7ebeab806a28cf052f365ecc2a47214a/lib/Module/ScanDeps.pm#L14-L18

**Comment by rschupp** [comment]
> A simple solution is to check if the key exists before entering either of the loops.

Go for it 😄 

> Using the same approach in the grep condition at

I don't understand what you mean here.

> Lifting the lc($module) and lc($used_by) calls to lexical vars outside the loops 

There's probably little gain here and it makes the code even harder to understand.

On File::Spec->case_tolerant()  vs  is_insensitive_fs I suggest
```
use constant is_insensitive_fs = File::Spec->case_tolerant() 
```
and replace all calls to the latter with the former.

**Comment by shawnlaffan** [comment]
> > Using the same approach in the grep condition at
> I don't understand what you mean here.

My thinking was that it might be faster to check for an exact match before trying a case insensitive one.  However on further thought this will not save meaningful time since it does not short circuit the grep.

If core modules are deemed OK to add as dependencies then many, maybe all, of the grep conditions in boolean context can be replaced with calls to ```List::Util::any```.

Another option is to unconditionally add items to the ```$rv->{somename}{uses}``` and ```$rv->{somename}{used_by}``` arrays and then uniq-ify them just before returning the data to the user.  That might hit other edge cases, though.

**Comment by rschupp** [comment]
> If core modules are deemed OK to add as dependencies then many, maybe all, of the grep conditions in boolean context can be replaced with calls to List::Util::any.

Good idea, but then 'PREREQ_PM' in 'Makefile.PL' should be amended with 
```
'List::Util' => '1.33'
```
(first version where it was implemented, which puts it into perl 5.20 and up, but `List::Util` is a dual-life module).
Care to amend your #22?

**Comment by rschupp** [comment]
> Another option is to unconditionally add items to the $rv->{somename}{uses} and $rv->{somename}{used_by} arrays and then uniq-ify them just before returning the data to the user. That might hit other edge cases, though.

One could even use a better data structure than an array "internally" (and convert on return of `scan_deps` and siblings). e.g. a hash with `module` or `lc(module)` (if is_case_insensitive) as key and first recorded `module` as value. But I'm too concerned about edge cases.

**Comment by shawnlaffan** [comment]
> Care to amend your #22?

Will do.

> But I'm too concerned about edge cases.

Yes, such a change would need a lot of checking for edge cases before implementing.

**Comment by shawnlaffan** [comment]
Resolved by #22
